sessionagent/response.go

```diff
@@ -0,0 +1,168 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015-2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package sessionagent
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/snapcore/snapd/logger"
+)
+
+// ResponseType is the response type
+type ResponseType string
+
+// "there are three standard return types: Standard return value,
+// Background operation, Error", each returning a JSON object with the
+// following "type" field:
+const (
+	ResponseTypeSync  ResponseType = "sync"
+	ResponseTypeAsync ResponseType = "async"
+	ResponseTypeError ResponseType = "error"
+)
+
+// Response knows how to serve itself, and how to find itself
+type Response interface {
+	ServeHTTP(w http.ResponseWriter, r *http.Request)
+}
+
+type resp struct {
+	Status int          `json:"status-code"`
```

Comment:
is this in itself ever JSON marshalled?  are they used in the test for unmarshalling?

Reply:
It is marshalled in the snapd API, so I left it as is here.  For better or worse, it is currently used by snapd-glib in place of the HTTP status code:

https://github.com/snapcore/snapd-glib/blob/master/snapd-glib/requests/snapd-json.c#L196

We could omit it in this interface, but that might be better suited to a larger discussion of cleaning up the API.

Reply:
And it looks like snapd-glib behaves this way because the Go client library does the same: `Client.do` discards the `http.Response` without checking the error code: it only checks the json body to detect error responses.

Reply:
sorry, I was referring to resp vs respJSON, don't we marshal only the latter?

Reply:
@jhenstridge I am a bit confused about the snapd-glib reference, this will be mostly be talked to from snapd itself, no?

Reply:
Due to where you attached this comment, I initially thought you were asking specifically about the `resp.Status` member, rather than the whole `resp` struct.  So my comments were more about "why is the status code marshalled in the response body?".  That's why I was looking at the two main clients for the snapd REST API.

I agree that it is unlikely libsnapd-glib would talk to the user session agent, at least with the initial scope of functionality.